### PR TITLE
Change C1 and C2 LineFunctions to a single type derived from NLPModel

### PR DESCRIPTION
**Pull request by abelsiqueira**
Discussion started on #35 @dpo @vepiteski

### CODE REVIEW

**Comment by abelsiqueira** [comment]
Building this on top of stats would allow returning the counters as `solver_specific`

File: src/linesearch/line_functions.jl [hunk 1]
```diff
@@ -13,11 +13,27 @@ represents the function ϕ : R → R defined by
     ϕ(t) := f(x + td).
 """
 type LineFunction <: AbstractNLPModel
```
Comment (dpo) [inline src/linesearch/line_functions.jl]: `LineModel`?

Comment (abelsiqueira) [inline src/linesearch/line_functions.jl]: Opinions on `redirect!`?

Comment (dpo) [inline src/linesearch/line_functions.jl]: I find it hard to guess what `redirect!` does but I can't really think of a better name...

Comment (abelsiqueira) [inline src/linesearch/line_functions.jl]: Same feeling, given that `reset!` is taken.

**Pull request merged**